Song breakup

theJenix opened the pull request:
It's not finished yet (still tracking down a bug), but I changed a lot of code and wanted to get some eyes on it.

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/fanClub

Conflicts:
	SoundStream-test/src/com/lastcrusade/soundstream/net/message/PlayStatusMessageTest.java

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/fanClub

Conflicts:
	SoundStream-test/src/com/lastcrusade/soundstream/net/message/PlayStatusMessageTest.java

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/SoundStream

theJenix committed:
Merge branch 'master' of git://github.com/TheLastCrusade/SoundStream into song_breakup

Conflicts:
	SoundStream-test/src/com/lastcrusade/soundstream/net/message/PlayStatusMessageTest.java

Initial checkin of messenger refactor, to split messages to allow play control
commands to be transmitted/received even while transmitting song data.

Refactored: cleaned up unused imports and variables, and moved Messenger and
supporting classes in .net.wire

theJenix committed:
Fixes in support of song breakup code

dgreenhalgh commented:
Holy fuck

theJenix commented:
To clarify, I believe the code is done with a only a logic bug remaining and I don't anticipate a lot of change past this (for this issue).  Also, a lot of the changes are like...cleaned up imports, or commented out unused variables....not that big a deal.  But yea, Messenger is completely rewritten, with a lot of functionality broken out into separate classes.

theJenix commented:
Oh, one other thing...I'm going to write a bunch of tests for this tomorrow.  Don't pull it in w/o the tests please :grin:

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThread.java:
```diff
@@ -135,8 +132,10 @@ public void run() {
                 if (messageRecvd) {
-                    //dispatch the message to the 
-                    sendMessageToHandler(mmMessenger.getReceivedMessage(), remoteDevice.getAddress());
+                    for (IMessage message : mmMessenger.getReceivedMessages()) {
```
Why do all of these variable names start with mm? I know you didn't do that this time around, but I am curious

theJenix replied:
This code started as an Android example and I kept the convention as I modified it.  We should probably make it consistent with the rest of the system...I'll do that as a later pull request.

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -67,24 +72,37 @@ public int compare(QueueEntry lhs, QueueEntry rhs) {
-        qe.score     = messageNo * (TransferSongMessage.class.isAssignableFrom(message.getClass()) ? 100 : 1);
-        qe.message   = message;
+        qe.messageNo     = messageNo;
+        qe.score         = messageNo * (TransferSongMessage.class.isAssignableFrom(message.getClass()) ? 100 : 1);
```
Can we have a comment about what 100 and 1 mean here?

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -67,24 +72,37 @@ public int compare(QueueEntry lhs, QueueEntry rhs) {
+            } else {
+                //otherwise, we're done
+                Log.i(TAG, "Message " + qe.messageNo + " finished writing");
+//            int len = messenger.serializeMessage(qe.message);
```
remove these if they are unneeded

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/core/AComplexDataType.java:
```diff
@@ -0,0 +1,100 @@
+        ByteBuffer bb = ByteBuffer.allocate(SIZEOF_INTEGER);
+        bb.putInt(integer);
+        output.write(bb.array());
+        //
```
Again, remove if unneeded

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/core/AComplexDataType.java:
```diff
@@ -0,0 +1,100 @@
+        }
+        ByteBuffer bb = ByteBuffer.wrap(in);
+        return bb.getInt();
+        // int value = input.read();
```
Here too

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/message/LibraryMessage.java:
```diff
@@ -28,7 +28,7 @@
 
 public class LibraryMessage extends ADataMessage {
-	private final String TAG = LibraryMessage.class.getName();
+//	private final String TAG = LibraryMessage.class.getName();
```
Uncomment?

theJenix replied:
It's not currently used, and it was giving me an "unused variable" warning.

reidbaker replied:
I agree with @ejohnson44 lets leave these uncommented

ejohnson44 commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,154 @@
+    private IMessage instantiateMessage(String messageName) {
+        try {
+            return (IMessage) Class.forName(messageName).newInstance();
+//            if (obj instanceof IMessage) {
```
remove

theJenix replied:
:watch: me go.

theJenix committed:
Fixed bugs with song breakup code.  Now verified to work correctly, and transfer
files as expected.

Fixed MessengerTest and added test for WireSendInputStream class

theJenix committed:
Code cleanup: comments, deleted dead code.

Added InputBuffer class to clean up places where we need to write to a
ByteArrayOutputStream and then navigate the buffer like an InputStream

Added LogUtil to let us test if the Log class is in the classpath (important for
unit testing)

theJenix committed:
Tuned buffer sizes in Messenger

Added file data buffering in FileReceiver to prevent frequent filesystem writes

Optimized bulk read/write operations in custom streams (calling the single byte
operations is very slow)

theJenix committed:
added comments around message scoring and the writeOne method in
MessageThreadWriter
removed dead code

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -51,21 +59,19 @@ public void testDeserializeMessage() throws Exception {
-
-        ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
-
+        InputStream is = simulateSendAndReceive(baos);
```
Could we change this variable is may stand for input stream but it looks like _is_ (this _is_ a sentence)

theJenix replied:
Mos :de:f

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -79,32 +85,84 @@ public void testDeserializeMessageMultiple() throws Exception {
         String testMessage2 = "This is another test of the messaging system";
         appendMessage(className, testMessage,  baos);
         appendMessage(className, testMessage2, baos);
+        //FIXME: rewrite this so that it's one message per packet, but multiple packets in the input stream
```
Did you want to do this before it was pulled in?

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -79,32 +85,84 @@ public void testDeserializeMessageMultiple() throws Exception {
+     * @return
+     * @throws IOException 
+     */
+    private File getTempTestFile() throws IOException {
```
Could you move this to testing utils

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -79,32 +85,84 @@ public void testDeserializeMessageMultiple() throws Exception {
     @Test
+    public void testDeserializeFileMessage() throws Exception {
+
+        File tempFile = getTempTestFile();
```
Comment after line?

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -79,32 +85,84 @@ public void testDeserializeMessageMultiple() throws Exception {
+     * @param file
+     * @throws IOException 
+     */
+    private void assertFileEquals(File expected, File actual) throws IOException {
```
This should probably be the SHA1 customAssert

theJenix replied:
Done

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -79,32 +85,84 @@ public void testDeserializeMessageMultiple() throws Exception {
+        assertNotNull(message);
+        assertTrue(message instanceof FileMessage);
+        //different paths
+        assertFalse(tempFile.getCanonicalPath().equals(((FileMessage)message).getFilePath()));
```
Why would it matter that that file paths are different?

theJenix replied:
This is really a sanity check of the test code...to make sure it simulated sending a file across the wire as best as possible.  If it doesn't the rest of the test may not be valid as we may not be going thru the packetization code.

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -168,12 +226,54 @@ public void testSerializeMessageMultiple() throws Exception {
-            Messenger messenger) throws IOException {
+            InputStream is) throws IOException {
+        return is;
+//        ByteArrayOutputStream baos = new ByteArrayOutputStream();
```
Remove these?

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -168,12 +226,54 @@ public void testSerializeMessageMultiple() throws Exception {
+    private InputStream simulateSendAndReceive(ByteArrayOutputStream baos) throws IOException {
+        byte[] bytes = baos.toByteArray();
+        baos.reset();
+        int testMessageNo = 1;
```
Why is this hardcoded? and why is it 1 not 0?

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/message/MessengerTest.java:
```diff
@@ -168,12 +226,54 @@ public void testSerializeMessageMultiple() throws Exception {
+        
+        format.serialize(baos);
+        FileReceiver fileFormat = new FileReceiver(message, null);
+        ByteBuffer bb = ByteBuffer.allocate(4);
```
4 units?

theJenix replied:
Addressed

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/wire/WireSendInputStreamTest.java:
```diff
@@ -0,0 +1,408 @@
+     * @return
+     * @throws IOException 
+     */
+    private File getTempTestFile(int size) throws IOException {
```
Can you move this to the utils?

theJenix replied:
:sweet_potato:

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/wire/WireSendInputStreamTest.java:
```diff
@@ -0,0 +1,408 @@
+    }
+
+    /**
+     * @param is
```
Reminder to rename _is_

theJenix replied:
Ok

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/net/wire/WireSendInputStreamTest.java:
```diff
@@ -0,0 +1,408 @@
+     */
+    @Test
+    public void testReadByteArrayIntIntOddPacketSize() throws IOException {
+//        int packetSize = 100;
```
Remove?

theJenix replied:
I have it commented out right now because I need to assess if its needed...let's leave for now (I added a TODO to this effect).

reidbaker commented on SoundStream-test/src/com/lastcrusade/soundstream/util/InputBufferTest.java:
```diff
@@ -0,0 +1,53 @@
+        assertEquals('c', bytes[1]);
+    }
+    
+    //TODO: need to add a test with -1s in the stream, to make sure we read things in properly
```
Does this need to happen before we pull the code in? If not is there a trello card

theJenix replied:
It's done

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -44,12 +42,29 @@
  */
 public class MessageThreadWriter {
 
+    /**
```
Empty doc strings. Either remove or fill in

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -67,24 +82,75 @@ public int compare(QueueEntry lhs, QueueEntry rhs) {
+     */
+    private int computeMessageScore(int messageNo, IMessage message) {
+        //transfer song has a higher score, so it is a lower priority message
+        return messageNo * (TransferSongMessage.class.isAssignableFrom(
```
multiple line ternary operator :thumbsdown: Is there a cleaner way to do this?

theJenix replied:
Yep...gotcha covered.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/MessageThreadWriter.java:
```diff
@@ -67,24 +82,75 @@ public int compare(QueueEntry lhs, QueueEntry rhs) {
-            messenger.reset();
-            Log.i(TAG, "Message " + qe.messageNo + " finished writing");
+            int read = qe.messageStream.read(outBytes);
+            if (LogUtil.isLogAvailable()) {
```
Might suggest making these log messages verbose

theJenix replied:
I made most of them debug, except the "message finished:' message...and I added a "message enqueued" info message.  This way we can keep close eye on messages queued and finished, while still getting easy access to the individual writes.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/core/AComplexDataType.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    /**
+     * @param nextBytes
```
These don't match your inputs

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/core/AComplexDataType.java:
```diff
@@ -0,0 +1,88 @@
+        if (length <= 0) {
+            throw new RuntimeException("Length is invalid: " + length);
+        }
+        // TODO: protection for laaaaaarge byte lengths
```
Does this need to happen before we pull this in? If not is there a trello card.

theJenix replied:
There is a trello card.  It doesnt need to happen before we pull in

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/core/ISerializable.java:
```diff
@@ -0,0 +1,48 @@
+     * @throws IOException 
+     * @throws MessageNotCompleteException 
+     */
+    public void deserialize(InputStream  input) throws IOException, MessageNotCompleteException;
```
if you are going to line up input and output line up IOException

theJenix replied:
:smile:

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/core/IntegerLength.java:
```diff
@@ -0,0 +1,62 @@
+     * 
+     */
+    public IntegerLength() {
+        // TODO Auto-generated constructor stub
```
Remove or do something

theJenix replied:
:+1:

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+
+/**
+ * A format for all messages.  All messages consists of:
+ *  4 byte length of the message, not including the length (using java.nio.ByteBuffer)
```
Change 4 to SIZEOF_INTEGER

theJenix replied:
:fish:

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+ *  4 byte length of the message, not including the length (using java.nio.ByteBuffer)
+ *  4 byte messenger version
+ *  class name (in bytes)
+ *  \n character to terminate the class name
```
change to END_OF_CLASS_CHAR? It might be nice to know that it is a \n though. What are your thoughts?

theJenix replied:
I think this is ok because I define a constant in this class for \n, and this is more interesting to folks who might be looking at the "on the wire" message, than people attempting to use the class.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+    private IMessage message;
+    private int length;
+
+    /**
```
Remove

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+    @Override
+    public void deserialize(InputStream input) throws IOException, MessageNotCompleteException {
+        //we assume that there is at least a length in the input stream
+        if (this.length == 0) {
```
This assumes that length is set to zero by default initialization. Lets set Length to zero above.

theJenix replied:
check

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+    private static final char END_OF_CLASS_CHAR = '\n';
+
+    private IMessage message;
+    private int length;
```
set to 0 here

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+            if (LogUtil.isLogAvailable()) {
+                Log.wtf(TAG, "Received message '" + messageName + "', but it does not implement IMessage");
+            }
+            throw new RuntimeException("Unable to instantiate message...this is a critical error.", ex);
```
Why the ... and this should probably log as well.

reidbaker replied:
log and or send us an email

theJenix replied:
Trello card to handle this...that would be future functionality.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/MessageFormat.java:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * @param input
+     * @param endChar
+     * @return
```
blank return

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/message/PauseMessage.java:
```diff
@@ -26,7 +26,7 @@
 
 public class PauseMessage implements IMessage {
-	private final String TAG = PauseMessage.class.getName();
+//	private final String TAG = PauseMessage.class.getName();
```
uncomment here and other places we have TAG

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    private static final int MIN_BYTES_READ_IN = 102400;
+
+    private static final int READ_BUFFER_SIZE = 1024;
```
units?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+            fileBuffer.write(byteBuffer, 0, read);
+            this.fileBytesLeft -= read;
+        }
+//        if (LogUtil.isLogEnabled()) {
```
Remove?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+
+        boolean readComplete = isInFileComplete();
+        
+        if (isBufferFilled() || readComplete) {
```
How do we get in to a situation where the buffer is not filled but the read is complete?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+
+    private boolean isBufferFilled() {
+        int available = fileBuffer.size();
+        return available > 0 && available >= Math.min(MIN_BYTES_READ_IN, this.fileBytesLeft);
```
Can we surround the return in parenthesis

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    private void flushBuffer() throws IOException {
+        Log.d(TAG, "Writing " + fileBuffer.size() + " bytes to incoming file (" + this.fileBytesLeft + " bytes left)");
```
This is one of the few places you are not checking if you can log before logging.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/FileReceiver.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    private File createRandomTempFile() throws IOException {
+        String filePrefix = UUID.randomUUID().toString().replace("-", "");
+        String extension = ".dat";
```
I think this file extension is different than what you use in your tests. Will you double check

theJenix replied:
It is, and that's on purpose/doesnt really matter.

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/MessageNotCompleteException.java:
```diff
@@ -0,0 +1,30 @@
+package com.lastcrusade.soundstream.net.wire;
+
+/**
+ * This exception is thrown to indicate that a full message has not been
```
Add this is an Expected exception

theJenix replied:
Gotcha

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+ */
+public class Messenger {
+
+//    private static final String TAG = Messenger.class.getSimpleName();
```
Reminder: uncomment

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+    @SuppressLint("UseSparseArrays")
+    private Map<Integer, WireRecvOutputStream> activeTransfers = new HashMap<Integer, WireRecvOutputStream>();
+
+    private List<IMessage>                    receivedMessages = new LinkedList<IMessage>();
```
not lined up but a ton of whitespace

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+     * Maximum size in bytes to write to a socket at a time.
+     * 
+     */
+   // private static final int MAX_WRITE_SIZE_BYTES = 1024;
```
Remove?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+        MessageFormat format = new MessageFormat(message);
+        InputBuffer buffer = new InputBuffer();
+        format.serialize(buffer);
+        // if this is a file message, open the file and prepare it for the write
```
rmove space before if

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+     */
+    private boolean processAndConsumePacket() throws IOException {
+        boolean received = false;
+        //REVIEW: character encoding issues may arise, but since we're controlling the class names
```
trello card?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+                this.receivedMessages.add(transfer.getReceivedMessage());
+            }
+        } catch (MessageNotCompleteException ex) {
+            //fall thru
```
Change to `//Expected exception when not done with message fall through`

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+            }
+        } catch (MessageNotCompleteException ex) {
+            //fall thru
+        } finally {
```
Remove finally block if not used?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/Messenger.java:
```diff
@@ -0,0 +1,247 @@
+        }
+    }
+
+    ///TODO
```
Do what?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/PacketFormat.java:
```diff
@@ -0,0 +1,111 @@
+ * transferring long file data over a single communications channel.
+ * 
+ * A packet consists of:
+ *  4 byte length of the message, not including the length (using java.nio.ByteBuffer)
```
Change 4 to SIZEOF_INTEGER

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/WireRecvOutputStream.java:
```diff
@@ -0,0 +1,144 @@
+        boolean received = false;
+        
+        try {
+            InputStream is = buffer.getInputStream();
```
Reminder to change is to something more descriptive

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/WireSendInputStream.java:
```diff
@@ -0,0 +1,211 @@
+        int read   = 0;
+        int toRead = Math.min(this.available(), maxLen);
+        while (availableInPacket() <= toRead && available() > 0) {
+//            int 
```
remove

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/net/wire/WireSendInputStream.java:
```diff
@@ -0,0 +1,211 @@
+            toRead      -= partial;
+        }
+
+//        packetIndex += toRead;
```
remove block?

reidbaker commented on SoundStream/src/com/lastcrusade/soundstream/service/ConnectionService.java:
```diff
@@ -351,7 +356,7 @@ public void onDisconnected() {
     }
 
-    public void findNewGuests() {
+    public void findNewGuests()  {
```
extra space?

theJenix committed:
Fixed bug with input streams containing multiple packets...now, Messenger will
process all available packets before asking for more data

Fixed MessengerTest so that all tests run and this case is covered.

Added to WireSendInputStreamTest

theJenix committed:
Removed fixme with issue about specific buffer sizes and file sizes

Restored 4k "optimal" buffer size

theJenix committed:
Uncommented unused TAG variable, and added  @SuppressWarnings("unused") to
remove the warning.  We've decided we would rather TAG be always available.

theJenix committed:
renamed input streams in test to be more descriptive
fixed message and packet format to be more correct (referencing SIZEOF_INTEGER)

theJenix committed:
Removed assertFileEquals in lieu of CustomAssert method of checking checksum
renamed verifyChecksum to assertChecksumsMatch

added Tag back to playmessage

theJenix committed:
Pulled message test routines to a util file

theJenix committed:
Cleaned up test variable names

added additional test to inputbuffertest to test -1 bytes

clarified code in computeMessageScore in MessageThreadWriter

added/cleaned up log messages in MessagethreadWriter

theJenix committed:
Cleanup of code, comments.

Added comments to key methods

theJenix committed:
Cleaned up dead code, and extra space

renamed inputstream from is to input

reidbaker commented:
:green_heart: Love it :grey_exclamation: lets pull this in
